File: frontend/src/App.test.ts
import { test, expect, vi, beforeEach, describe } from 'vitest';
import { render, screen } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import App from './App.svelte';
import Header from './Header.svelte';
import CommentPane from './CommentPane.svelte';

/*
In order to understand how to write the tests, first we looked at the lab slides, then we had to do some reading from svelte testing documentation and npm documentation. We also read up on documentation in NYT's response fields to help make tests on articles.
Here are the links of the documentation that we used. The tests were succesfully ran on our machines. Check test_proof for screenshots on the passing tests.
https://docs.npmjs.com/downloading-and-installing-node-js-and-npm - Node and npm documentation 
https://svelte.dev/docs/svelte/testing - Svelte testing documentation
https://svelte.dev/tutorial/svelte/welcome-to-svelte - Intro Svelte code documentation
https://canvas.ucdavis.edu/courses/993295/files/folder/Lab%20Materials/week%205? - Week 5 Lab Slides
https://developer.nytimes.com/docs/articlesearch-product/1/overview#response-fields - NYT's API Response Fields
*/

// The code down below is to return a fake API key and return a fake article so that we do not need to get the backend. 
// The fake article has components such as the word_count, multimedia from real different articles however the headline and abstract are fake and tested in the code. 
vi.stubGlobal("fetch", vi.fn((url: string) => {
    if (url === "/api/key") {
        return Promise.resolve({
            ok: true,
            json: () => Promise.resolve({
                apiKey: "fake-key"
            }),
        });
    }
    if (url == "/api/me") {
        return Promise.resolve({
            ok: true,
            json: () => Promise.resolve({
                name: "Moderator", 
                email: "moderator@hw3.com",
                moderator: true
            }),
        });
    }
    if (url.includes("/api/comments")) {
        return Promise.resolve({
            ok:true,
            json: () => Promise.resolve([
                { _id:'1', parent_id:'root', text:'Top-level comment', user:'user@hw3.com', moderated:false },
                { _id:'2', parent_id:'1', text:'Reply comment', user:'user@hw3.com', moderated:false },
                { _id:'3', parent_id:'root', text:'COMMENT REMOVED BY MODERATOR!', user:'moderator@hw3.com', moderated:true }
            ])
        });
    }
    if (url === "/api/news") {
        return Promise.resolve({
            ok: true,
            json: () => Promise.resolve({
                response: {
                    docs: [
                        {
                            abstract: "Fake article about UC Davis and Sacramento",
                            headline: {main: "UC Davis and Sacramento Club"},
                            word_count: "500",
                            byline: {original: "By Navjeet Hundal"},
                            multimedia: {
                                default: {
                                    "url": "https://static01.nyt.com/images/2023/05/04/multimedia/04nat-davis-arrest-01-bktg/04nat-davis-arrest-01-bktg-articleLarge.jpg",
                                    "height": 399,
                                    "width": 600
                                },
                            },
                        },
                    ],
                },
            }),
        });
    }
    return Promise.reject(new Error('unhandled fetch ${url}'));
}));

vi.mock('./CommentPane.svelte', () => ({
  default: () => '<div data-testid="mock-commentpane"></div>'
}));

describe('Layout and header', () => {
  // This test is to test out if today's date is displayed in the Header and in the right format. It searchs the text to find the formatted date on the page.
  test('todays date', () => {
    render(Header, {
      props:{ userInfo:{ email:'guest@ucdavis.edu', name:'Guest', moderator:false },
              updateUserInfo: () => {} }
    });

    const today = new Date().toLocaleDateString('en-US', { weekday:'long', month:'long', day:'numeric', year:'numeric' });
    expect(screen.getByText(today)).toBeInTheDocument();
  });
 
  // This test is to test out if fake articles can be displayed in the UI. We utilize the fake article created above. 
  // The way the code works is that we render the app, then we wait until the fake article headline appears in the DOM. 
  // Then we check that it exists. Then we find the article’s abstract using a regular expression match. Then we check if that exists as well.  
  test('fake content', async () => {
    render(App);
    expect(await screen.findByText('UC Davis and Sacramento Club')).toBeInTheDocument();
    expect(screen.getByText(/Fake article about UC Davis/i)).toBeInTheDocument();
  });
});

vi.unmock('./CommentPane.svelte');  

const commentProps = {
  visibility : { state: true },         
  data       : {
    articleId:'nyt-fake',
    comments : [                       
      { id:'1', parent: 'root', content:'Top-level comment', username:'user@hw3.com',
        date:new Date().toISOString(), locale:'USA', replyNum:0 },

      { id:'2', parent: '1', content:'Reply comment', username:'user@hw3.com',
        date:new Date().toISOString(), locale:'USA', replyNum:0 },

      { id:'3', parent: 'root', content:'COMMENT REMOVED BY MODERATOR!', username:'mod@hw3.com',
        date:new Date().toISOString(), locale:'USA', replyNum:0 }
    ]
  },
  updateCommentsHandler: () => {},
  userInfo : { email:'user@hw3.com', name:'User', moderator:false }
};

describe('CommentPane rendering', () => {
  beforeEach(() => vi.useRealTimers());   
  
  // This test is to see if the comments are showing up under articles. 
  // The way the code works is we render commentpane, and then wait until we see the comment show up and then we check if its exists. 
  test('shows top-level comment', async () => {
    render(CommentPane, { props: commentProps });
    expect(await screen.findByText('Top-level comment')).toBeInTheDocument();
  });
  
  // This test is to see if the comments are showing up and that replys are working correctly.
  test('reply to comment', async () => {
    render(CommentPane, { props: commentProps });
    expect(await screen.findByText('Reply comment')).toBeInTheDocument();
  });

  // This test is to see if the deleted comment is working corretly. 
  // The code works by rendering the commentpane and then finding the comment removed by moderator text which should show up when deleting a comment.
  test('shows deleted comment', async () => {
    render(CommentPane, { props: commentProps });
    expect(await screen.findByText('COMMENT REMOVED BY MODERATOR!')).toBeInTheDocument();
  });
});

File: backend/app_test.py
import pytest
from app import app as create_app

# In order to understand how to write the tests, first we looked at the lab slides, then we had to do some reading from pytest documentation and flask documentation. We also read up on documentation in NYT's response fields to help make tests on articles.
# Here are the links of the documentation that we used. 
# https://docs.pytest.org/en/stable/example/simple.html - Pytest documentation
# https://flask.palletsprojects.com/en/latest/testing/#the-test-client - Flask documentation
# https://canvas.ucdavis.edu/courses/993295/files/folder/Lab%20Materials/week%205? - Week 5 Lab Slides
# https://developer.nytimes.com/docs/articlesearch-product/1/overview#response-fields - NYT's API Response Fields

# This is the test client.
@pytest.fixture
def client():
    create_app.config["TESTING"] = True
    with create_app.test_client() as client:
        yield client

# The code down below is to test the API key. Specifically it is to make sure that the API key is returned.
# How the code works is that response sends a get request to the /api/key route. Then the server replies with code 200 which means everything is going ok.
# Next, we convert the JSON response into Python and check if the "data" has the apiKey in it. After we have found the apiKey we check to see that it is a string. 
# Finally, we make sure that the key is real by checking its length, if greater than 25 we deem the key to be real. We chose 25 for the length as the key we used for this project was roughly 30 characters long so a key must be somewhat similar length.
def test_api_key(client):
    response = client.get("/api/key")
    assert response.status_code == 200
    data = response.get_json()
    assert "apiKey" in data
    assert isinstance(data["apiKey"], str)
    assert len(data["apiKey"]) > 25

# The code down below is to test /api/news route and structure and content. Specifically it is to check if it returns articles and that the article contains either "UC Davis" or "Sacramento".
# How the code works is that response sends a get request to the /api/news route. Then the server replies with code 200 which means everything is going ok.
# Next, we convert the JSON response into Python. According to the NYT API Response Field, the docs is the list of article results so we end up checking that. 
# Finally, if there are articles which there should be then we go through each article to check if there is "UC Davis" or "Sacramento" in the article.
def test_news_content(client):
    response = client.get("/api/news")
    assert response.status_code == 200
    data = response.get_json()
    assert "response" in data
    assert "docs" in data["response"]
    articles = data["response"]["docs"]
    assert isinstance(articles,list)
    
    if not articles:
        pytest.skip("Zero articles returned")
        
    for content in articles:
        abstract = (content.get("abstract")).lower()
        if "uc davis" in abstract or "sacramento" in abstract:
            return 
    assert False, "No UC Davis or Sacramento in the articles"

# The code down below is to test the /api/comments/<comment_id>/moderater route.
# First it will create a comment, and then deleting it using the moderator permission. 
# The code works by first making a POST to /api/comments to simulate a comment. THen a POST sent to /api/comments/<comment_id>/moderater to simulate deleting a comment.
# Then it should show that the comment was deleted.          
def test_moderator_delete(client):
    article_id = "pytest-article"
    post = client.post("/api/comments", json={
        "article_id": article_id,
        "text": "Delete me please",
        "parent_id": "root"
    }, headers={"X-User-Email": "pytest@test.com", "X-User-Is-Moderator": "true"})
    comment = post.get_json()
    comment_id = comment["_id"]
    
    res = client.post(f"/api/comments/{comment_id}/moderater", headers={
        "X-User-Email": "pytest@test,com",
        "X-User-Is-Moderator": "true"
    })
    assert res.status_code == 200
    message = res.get_json()
    assert message["message"] == "Comment deleted"  

# The code down below is to test /api/comments route. 
# The code should create a comment and then retrive that comment for the same article.    
def test_get_comments(client):
    article_id = "pytest-article"
    post = client.post("/api/comments", json={
        "article_id": article_id,
        "text": "Comments",
        "parent_id": "root"
    }, headers={"X-User-Email": "pytest@test.com", "X-User-Is-Moderator": "true"})
    
    res = client.get(f"/api/comments?article_id={article_id}")
    assert res.status_code == 200
    comments = res.get_json()
    assert isinstance(comments, list)
    assert any(c["text"] == "Comments" for c in comments)

# The code down below is to test if we can recover a saved article. 
# The code works by saving an article via POST, then fetches it using GET to verify the data inside. 
def test_get_article(client):
    article_id = "https://www.nytimes.com/2023/05/04/us/davis-stabbings-aest-california.html"
    client.post("/api/articles", json={
        "url": article_id,
        "headline": "Article",
        "abstract": "Summary",
        "byline": "Author"
    })

    res = client.get(f"/api/articles/{article_id}")
    assert res.status_code == 200
    article = res.get_json()
    assert article["url"] == article_id
    assert article["headline"] == "Article"
    
# The code down below is to test if a non-moderator cannot delete a comment.
# The code posts a comment and then tries to delete it with a non-moderator, which should not work because only moderators are able to do that.
def test_non_moderator_delete(client):
    post = client.post("/api/comments", json={
        "article_id": "dummy test",
        "text": "No delete",
        "parent_id": "root"
    }, headers={"X-User-Email": "pytest@test.com", "X-User-Is-Moderator": "false"})
    
    comment_id = post.get_json()["_id"]
    res = client.post(f"/api/comments/{comment_id}/moderater", headers={
        "X-User-Is-Moderator": "false"
    })
    assert res.status_code == 400
    assert res.get_json()["error"] == "Unauthorized/Not Moderator"

      
    

